### PR TITLE
Speedup rendering functions

### PR DESCRIPTION
**Pull request by carlo-bramini**
See issue #571

### CODE REVIEW

File: src/synth/fluid_synth.c [hunk 1]
```diff
@@ -3816,26 +3816,52 @@ fluid_synth_write_float(fluid_synth_t *synth, int len,
...
+        }
+        while (++n < 0);
     }
+    while (len);
 
-    synth->cur = l;
+    synth->cur = cur;
 
     time = fluid_utime() - time;
     cpu_load = 0.5 * (fluid_atomic_float_get(&synth->cpu_load) + time * synth->sample_rate / len / 10000.0);
```
Comment (derselbst) [inline src/synth/fluid_synth.c]: When control reaches here, `len` will be zero, causing a DIV by zero. Recompile with `cmake -Denable-trap-on-fpe=1` to get a SIGFPE.

Comment (carlo-bramini) [reply src/synth/fluid_synth.c]: Argh, I did not notice the use of len into cpu_load, I will fix it.

**Comment by jjceresa** [comment]
Thanks for speed enhancement, however this new implementation doesn't behave as before. 
When  `fluid_synth_write_xxx()` is called with `len` to 0, the function will still call fluid_synth_render_blocks() and write in caller supplied buffer here:

*left_out = (float) left_in[n];
*right_out = (float) right_in[n];

This could lead with possible memory violation access. To fix that  the outer loop `do; {..} while(size)`
should be `while(size) {...}`.
***
Please, why did you use a reverse indexing ?. Shouldn't a classic indexing be more faster ? and easier to read, like this:
for (i=0,  left_in  += cur,   right_in += cur;   i < n;   i++)
{
            *left_out = (float) left_in[i];
            *right_out = (float) right_in[i];

            left_out += lincr;
            right_out += rincr;
}

**Comment by jjceresa** [comment]
Thanks for speed enhancement, however this new implementation doesn't behave as before.
When `fluid_synth_write_xxx()` is called with `len` to 0, the function will still call fluid_synth_render_blocks() and write in caller supplied buffer here:

*left_out = (float) left_in[n];
*right_out = (float) right_in[n];

This could lead with possible memory violation access. To fix that the outer loop do; {..} while(size)
should be while(size) {...}.
****
Please, why did you use a reverse indexing ?. Shouldn't a classic indexing be more faster ? and easier to read, like this:
'''
for (i=0, left_in += cur, right_in += cur; i < n; i++)
{
    *left_out = (float) left_in[i];
    *right_out = (float) right_in[i];

    left_out += lincr;
    right_out += rincr;
}
'''

**Comment by carlo-bramini** [comment]
I fixed the case when `len` is zero, thank you.

> Shouldn't a classic indexing be more faster

No, I don't think so. The readability of the code is quite subjective, but in my opinion the condition `(++n < 0)` is typically evaluated faster than `(++i < n)`.

**Comment by derselbst** [comment]
> > Shouldn't a classic indexing be more faster
> 
> No, I don't think so.

Reversed indexing is fine for me. 

> condition (++n < 0) is typically evaluated faster than (++i < n).

I wouldn't nit-pick too much on those issues. After all, we are only talking about integers, and most of the time will be spent when accessing `left_in` and `right_in` and loading the data into the cache. That said, what you could try is traversing `left_in` and `right_in` in separate loops, to avoid potential cache conflicts (https://stackoverflow.com/q/8547778). And declaring `left_out`, `right_out`, `left_in` and `right_in` as `FLUID_RESTRICT`.

**Comment by jjceresa** [comment]
>That said, what you could try is traversing left_in and right_in in separate loops, to avoid potential cache conflicts.

1)Interesting idea !.

2) Also to simplify a bit we could  replace `do; {..} while(size)` by  `while(size) {...}` this will allow to remove `if (FLUID_LIKELY((size = len) > 0))`

**Comment by carlo-bramini** [comment]
> 1. Also to simplify a bit we could  replace `do; {..} while(size)` by  `while(size) {...}` this will allow to remove `if (FLUID_LIKELY((size = len) > 0))`

Actually, I did that intentionally instead of using a while(), it is true that the compiler could unroll the loop, but in my opinion it would be worth to help on this piece of code.

**Comment by carlo-bramini** [comment]
> what you could try is traversing left_in and right_in in separate loops, to avoid potential cache conflicts

It is a good idea, although this seems to be platform dependant.
Actually, my platform performs better with a single loop rather than two splitted loops.

**Comment by jjceresa** [comment]
>On my plaftorms, the improvements were very interesting.
 The improvement comes out by eliminating the execution costs of the test at every cycle.

I'm curious of the improvements you got. Would you please report these result ?

**Comment by derselbst** [comment]
> Actually, my platform performs better with a single loop rather than two splitted loops.

Ok.

> > On my plaftorms, the improvements were very interesting.
> > The improvement comes out by eliminating the execution costs of the test at every cycle.
> 
> I'm curious of the improvements you got. Would you please report these result ?

Me too.

Unit test passed, will merge it now, because it's convenient for me. Thanks.

**Pull request merged**